src/drivers/intel/cavs/mn.c

```diff
@@ -102,19 +148,53 @@ int mn_set_mclk(uint16_t mclk_id, uint32_t mclk_rate)
 		break;
 	default:
 		trace_mn_error("error: invalid mdivr_val %d", mdivr_val);
-		ret = -EINVAL;
-		goto out;
+		return -EINVAL;
 	}
 
-	mn_reg_write(MN_MDIVCTRL, mdivc);
 	mn_reg_write(MN_MDIVR(mclk_id), mdivr);
+	return 0;
+}
+
+int mn_set_mclk(uint16_t mclk_id, uint32_t mclk_rate)
+{
+	int ret = 0;
+
+	if (mclk_id >= DAI_NUM_SSP_MCLK) {
+		trace_mn_error("error: mclk ID (%d) >= %d",
+			       mclk_id, DAI_NUM_SSP_MCLK);
+		return -EINVAL;
+	}
+
+	spin_lock(mn_lock);
+
+	mclk_sources_used[mclk_id] = false;
+
+	if (is_mclk_source_in_use())
+		ret = check_current_mclk_source(mclk_rate);
+	else
+		ret = setup_initial_mclk_source(mclk_rate);
+
+	if (ret < 0)
+		goto out;
+
+	mclk_sources_used[mclk_id] = true;
+
+	ret = set_mclk_divider(mclk_id,
+			       ssp_freq[mclk_source_clock].freq / mclk_rate);
```

Comment:
@jajanusz just to clear my confusion: This means that the last SSP port to be configured determines the mdivr, right? Is the expectation that all the mclk rates requested in topology should be the same for all SSP ports?

Reply:
Unfortunetly yes, if you configure MCLK0 for SSP0 and then also MCLK0 for SSP1 then you gonna overwrite first config, it was like that all the time, idk if assumption here was that it's driver/topology responsibility to know what it does. Only improvement IRT to that in this PR is that now you are not gonna overwrite source for MCLK0 when playing with MCLK1. We can add these checks tho.

Reply:
OK thanks for clarifying!